Piwik: adding custom variable support

harrietgrace opened the pull request:
Updating Piwik to let people send custom variables. Sticking with the format that was requested by [this ticket](https://segment.zendesk.com/agent/tickets/22930):

```
{ cvar: {
    1:["UserId","898032"],
    2:["Things",""],
    3:["Names","Cool Beans"]
   }
 }
```

Sending it in this format on every call gives us two important features - a) you can specify particular indexes for each variable (something that piwik allows, and needs to be sent with every setCustomVariable call), and b) the ability to send different variables on each call which is something that we wouldn't get if we were getting users to add them through their dashboard settings. 

Piwik docs are [here](http://developer.piwik.org/guides/tracking-javascript-guide) and the custom vars section is about 1/3 of the way down the page.

@ndhoule

Commit:
Piwik: adding custom variable support

Commit:
Adding extra tests, support for passing arrays

ndhoule commented:
I think you refactored how to pass custom vars—how does that work now?

ndhoule commented on lib/piwik/index.js:
```diff
@@ -64,8 +67,35 @@ Piwik.prototype.track = function(track){
+  //var customVariableLimit = this.options.customVariableLimit;
+
+  // Not currently enabling setting your own variable limit, setting this to the default
+  var customVariableLimit = 5;
```
Can you turn this into a GH ticket and remove these TODOs?

ndhoule commented on lib/piwik/index.js:
```diff
@@ -64,8 +67,35 @@ Piwik.prototype.track = function(track){
+  var formattedVars = {};
+
+  if (is.array(customVariables)) {
+    formattedVars = reduce(customVariables, function(formattedVars, value, i){
```
Break this out into a utility function rather than inlining it

harrietgrace commented:
You pass them in the context like this:
```
{ integrations: {
  Piwik: {
    customVars: {
      1: ["UserId","6116"],
      2: ['SubscriptionId', ''],
      3: ['PlanName', 'ENTERPRISE']
      }
    }
  }
}
```

ndhoule commented on lib/piwik/index.js:
```diff
@@ -64,8 +67,35 @@ Piwik.prototype.track = function(track){
+    }, {});
+  }
+
+  if (is.object(customVariables)) formattedVars = customVariables;
```
What happens when it's a non-array/non-object value?

Sidenote—does `is.object` return `true` for arrays?

harrietgrace replied:
It won't send anything if you provide them with the wrong type, there are a couple of tests in there for incorrect types. 

No, `is.object` returns false for arrays.

ndhoule commented on lib/piwik/index.js:
```diff
@@ -64,8 +67,35 @@ Piwik.prototype.track = function(track){
+
+  if (is.object(customVariables)) formattedVars = customVariables;
+
+  if (!is.null(formattedVars) && !is.undefined(formattedVars)) {
```
```
if (formattedVars) {
  // ...
}
```

ndhoule commented on lib/piwik/index.js:
```diff
@@ -64,8 +67,35 @@ Piwik.prototype.track = function(track){
+
+  if (!is.null(formattedVars) && !is.undefined(formattedVars)) {
+    each(formattedVars, function(key, valuePair){
+      if (parseInt(key) <= customVariableLimit) push('setCustomVariable', key, valuePair[0], valuePair[1], 'page');
```
Always pass a radix to `parseInt`: `parseInt(key, 10)`

ndhoule commented on lib/piwik/index.js:
```diff
@@ -64,8 +67,35 @@ Piwik.prototype.track = function(track){
+      if (parseInt(key) <= customVariableLimit) push('setCustomVariable', key, valuePair[0], valuePair[1], 'page');
+    });
+  }
+
```
You could simplify this logic if you converted objects into arrays, rather than the other way around:

```javascript
if (!is.object(customVariables) || !is.array(customVariables)) {
  customVariables = [];
}

if (is.object(customVariables)) {
  customVariables = objToArray(customVariables);
}

for (var i = 0; i < this.settings.customVariableLimit; i += 1){
  if (customVariables[i]) {
    push('setCustomVariable', i + 1, customVariableLimit[i][0], customVariableLimit[i][1], 'page');
  }
}
```

ndhoule replied:
(gotta test that, wrote it without running it ^)

Commit:
minor logic update and style changes, adding in new setting

harrietgrace commented:
Updated so that the final formatted variables are an array instead of an object. Also added in the setting for custom variable limits which will get added to the metadata under advanced settings - if they don't put anything it will default to 5 (which is Piwik's default). Added some extra tests as well.

ndhoule commented on lib/piwik/index.js:
```diff
@@ -15,6 +17,7 @@ var Piwik = module.exports = integration('Piwik')
   .global('_paq')
   .option('url', null)
   .option('siteId', '')
+  .option('customVariableLimit', '')
```
Set this thang's default as `5`

ndhoule commented on lib/piwik/index.js:
```diff
@@ -64,8 +67,42 @@ Piwik.prototype.track = function(track){
   var value = track.value() || track.revenue();
 
+  var options = track.options('Piwik');
+  var customVariableLimit = this.options.customVariableLimit || 5;
```
If you set the default above, you can eliminate the `||` here

ndhoule commented on lib/piwik/index.js:
```diff
@@ -64,8 +67,42 @@ Piwik.prototype.track = function(track){
+    result.push(value);
+    return result;
+  });
+};
```
eof newline

ndhoule commented on lib/piwik/test.js:
```diff
@@ -119,6 +120,138 @@ describe('Piwik', function(){
+      });
     });
   });
 });
```
Can you add an eof newline here, too?

harrietgrace replied:
Damn, I must have accidentally turned off that sublime extension you showed me for that

ndhoule commented on lib/piwik/test.js:
```diff
@@ -34,6 +34,7 @@ describe('Piwik', function(){
       .global('_paq')
       .option('siteId', '')
       .option('url', null)
+      .option('customVariableLimit', '')
```
Same default thing here

ndhoule commented on lib/piwik/index.js:
```diff
@@ -64,8 +67,42 @@ Piwik.prototype.track = function(track){
+
+Piwik.prototype._formatToArray = function(object){
+  return reduce(object, [], function(result, key, value){
+    key = parseInt(key, 10);
```
We aren't using the key here, just parsing it. Also, we aren't guaranteed to store the value at the index given by `key` because order of iteration over objects isn't guaranteed as part of the ES spec*, so you need to parse and use the key here. It also won't filter out non-numeric keys. Check out my prior comment for the fix.

Run this code for a repro:

```js
var obj = {};
obj.a = 'shirt'
obj['1'] = 'pants';
obj['0'] = 'shoes';
for (var key in obj) { console.log(key); }
```

* Iteration order behavior is platform specific: https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Statements/for...in

Commit:
Adding defaults, more cleaning

ndhoule commented on lib/piwik/index.js:
```diff
@@ -6,6 +6,8 @@
 var push = require('global-queue')('_paq');
 var each = require('each');
+var is = require('is');
+var reduce = require('ianstormtaylor/reduce');
```
You can remove this now, since we're not using it anymore

ndhoule commented on lib/piwik/test.js:
```diff
@@ -119,6 +120,119 @@ describe('Piwik', function(){
+        analytics.called(window._paq.push, ['setCustomVariable', '3', 'PlanName', 'ENTERPRISE', 'page']);
+      });
+
+      it('should send multiple custom variables that get passed as an object', function(){
```
language on this is outdated now—should mirror above, `should send multiple custom variables that get passed as customVars`. You could also break the object you're passing into a variable to make later easier

ndhoule commented on lib/piwik/test.js:
```diff
@@ -119,6 +120,119 @@ describe('Piwik', function(){
+        analytics.didNotCall(window._paq.push, ['setCustomVariable', '7', 'ThisIsntAThing', 'Apples', 'page']);
+      });
+
+      it('should pick up the max variable setting change and send more than 5 variables', function(){
```
`should send up to 'customVariableLimit' number of variables`

ndhoule commented on lib/piwik/test.js:
```diff
@@ -119,6 +120,119 @@ describe('Piwik', function(){
+        analytics.didNotCall(window._paq.push, ['setCustomVariable', '1', 'UserId', '', 'page']);
+      });
+
+      it('shouldn\'t send more than the max number of custom variables', function(){
```
`should not send more than 'customVariableLimit' number of variables`

ndhoule commented:
lgtm otherwise
